reports.py

```diff
@@ -673,6 +680,22 @@ def plot_coverage(
             segment_depths.setdefault(row[0], []).append(float(row[2]))
             domain_max += 1
 
+    # Binning
+    bin_size = 1
+    if bin_large_plots:
+        # Number points plotted after binning is set to true width of plot (excluding
+        # axis titles and labels and whitespace around plot), calculated by trial and error
+        # There is one bin per pixel
+        preferred_domain = int(0.907634 * plot_width - 103.207)
```

Comment:
Are these constants a function of the DPI of the system? If so, can the width in pixels be calculated similar to how it's done below via `DPI = plot_dpi or fig.get_dpi()`? Matplotlib thinks more in relative dimensions than in pixels to suit its vector backends, so we generally have to take DPI into account as we're [doing below](https://github.com/broadinstitute/viral-ngs/blob/master/reports.py#L679-L680). Can the binning code block be moved within the `with plt.style.context(plot_style)` context below to take advantage of the sizing code there? Or if it's more a function of the data range than the dimensions of the plot (with axis labels, ticks, etc.), perhaps something like [`ax.get_xlim()`](https://matplotlib.org/2.0.0/api/_as_gen/matplotlib.axes.Axes.get_xlim.html) (using the `ax` below that references the subplot) can be used?

Reply:
> Are these constants a function of the DPI of the system?

No, I don't think so.

This is just the width in pixels of the actual plot part of the plot (not including the whitespace and text around the plot)—see attached. Not sure how to actually get that value in a less hacky way—if you happen to know, I would love to know too. I tried different plot_width values and checked what the inside plot width was and did a linear regression on the pairs—hence the constants (messy, I agree). If you prefer I can just use plot_width instead—it won't be the width in pixels of the actual plotted part of the plot (so we won't be able to say that there are _n_ bins per pixel), but it won't have ugly constants.

![Plasmodium_coverage_plot-01](https://user-images.githubusercontent.com/6245320/58912302-bbedc500-86e7-11e9-9d86-63c4b3cfd97a.png)

Reply:
And on second look that variable name sucks. I'll rename it to inner_plot_width and google for a better way to get that width....

Reply:
I think getting the exact size in pixels is hard because the renderer doesn't know how big the axes are going to be and it depends on the text of the y axis labels for example.

Reply:
[`ax.get_xlim()`](https://matplotlib.org/2.0.0/api/_as_gen/matplotlib.axes.Axes.get_xlim.html) should give the dimension of what you annotated in the image above, in arbitrary units.

Reply:
I agree. And I haven't any solutions online. I suppose even if I added plot_height to my little linear regression to capture the effect of font size, I still couldn't capture the additional width added by the y axis tick labels in any non-absurd way.... Well. Let's use plot_width instead of trying to get the width of the inside plot. It is not as pleasing but I think it is okay.

Reply:
Oops, sorry, didn't see your response @tomkinsc! ax.get_xlim() gives me (0.0, 1.0), but I do not yet know what that means will try to figure out what I can do with it.

Reply:
Ah, I see. @tomkinsc, `get_xlim` gives us the limits of the x axis; in this particular case it gives us -1.2e7 to 24.8e7 (more precisely, -1,181,769 to 24,817,161). What I am looking for is the width of that space in pixels. The purpose of getting the width of that space in pixels is to then be able to set the bin size so that there is some known, predetermined number of bins per pixel (currently 1 bin per pixel).

Reply:
Matplotlib doesn't think in pixels though, but physical units (inches), which can be converted to pixels based on the DPI of the system (so with a desired pixel dimension and the DPI of the system, you can [set the size in pixels](https://github.com/broadinstitute/viral-ngs/blob/master/reports.py#L679-L680)). Should be able to interrogate the current size, get inches, and multiply by DPI to get the size in pixels.

Reply:
Absolutely. I agree with you 100%. But what I would be looking for is not the width of the plot (`plot_width` in pixels or `plot_width / DPI` in inches), but the width of that smaller part of the plot I outlined in blue above. If you know how to interrogate the current size (in inches or pixels!) of that smaller part of the plot I outlined in blue above, then that would definitely be really cool and I would love to know how. As far as I can tell `get_xlim` does not do that, but I would not be sad to be proved wrong.

Reply:
If [this](https://stackoverflow.com/a/19306776) works, possibly:
```
bbox = ax.get_window_extent().transformed(fig.dpi_scale_trans.inverted())
width, height = bbox.width, bbox.height
width *= fig.dpi
height *= fig.dpi
```
The matplotlib API leaves much to be desired in terms of usability/memorability; half the time I end up inspecting objects with `dir()`...

Reply:
It works!!!! Thou art sorcery! Thank you so so much!!!!